contrib/win32/openssh/telemetry.c

```diff
@@ -0,0 +1,177 @@
+{
+    TraceLoggingRegister(g_hProvider1);
+    TraceLoggingWrite(
+        g_hProvider1,
+        "Key",
+        TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage),
+        TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
+        TraceLoggingString(key, "Status")
+    );
+    TraceLoggingUnregister(g_hProvider1);
+}
+
+void send_shell_telemetry(const int pty, const int shell_type)
+{
+    TraceLoggingRegister(g_hProvider1);
+    TraceLoggingWrite(
+        g_hProvider1,
+        "Shell",
+        TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage),
+        TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
+        TraceLoggingInt16(pty, "PTY"),
+        TraceLoggingInt16(shell_type, "Type")
+    );
+    TraceLoggingUnregister(g_hProvider1);
+}
+
+void send_sign_telemetry(const char* sign_status)
+{
+    TraceLoggingRegister(g_hProvider1);
+    TraceLoggingWrite(
+        g_hProvider1,
+        "Signing",
+        TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage),
+        TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
+        TraceLoggingString(sign_status, "Status")
+    );
+    TraceLoggingUnregister(g_hProvider1);
+}
+
+void send_ssh_telemetry(const char* conn)
+{
+    TraceLoggingRegister(g_hProvider1);
+    TraceLoggingWrite(
+        g_hProvider1,
+        "Connection",
+        TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage),
+        TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
+        TraceLoggingString(conn, "Status")
+    );
+    TraceLoggingUnregister(g_hProvider1);
+}
+
+void send_sshd_telemetry(const int num_auth_methods, const char** auth_methods,
+    const unsigned int num_ports, const int ports[])
+{
+    char* auth_buffer = NULL;
+
+
+    if (num_auth_methods == 0) {
+        auth_buffer = (char*)malloc(5 * sizeof(char));
+        strcpy_s(auth_buffer, 5, "none");
+    }
+    else {
+        // concatenate all the auth methods into a 
+        // single string to pass to tracelogging
+        size_t buffer_size = (size_t)num_auth_methods;
+        for (int i = 0; i < num_auth_methods; i++) {
+            buffer_size += strlen(auth_methods[i]);
+        }
+        auth_buffer = (char*)malloc(buffer_size * sizeof(char));
+        auth_buffer[0] = '\0';
+        for (int i = 0; i < num_auth_methods; i++) {
+            strcat_s(auth_buffer, buffer_size, auth_methods[i]);
+            if (i < num_auth_methods - 1) {
+                strcat_s(auth_buffer, buffer_size, ",");
+            }
+        }
+    }
+
+    TraceLoggingRegister(g_hProvider1);
+    TraceLoggingWrite(
+        g_hProvider1,
+        "SSHD",
+        TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage),
+        TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
+        TraceLoggingInt32Array(ports, num_ports, "Port"),
+        TraceLoggingString(auth_buffer, "Auth Methods")
+    );
+    TraceLoggingUnregister(g_hProvider1);
+    free(auth_buffer);
+}
+
+void send_startup_telemetry(const char* ssh_version, const char* peer_version, 
```

Comment:
Rename it to send_ssh_version_telemetry

Reply:
This is a common code for both client and server. See if you can add it's coming from client or server